### ISSUE TITLE
psycopg2 with python3.9

### ISSUE THREAD

**Issue by fossildoc** [qgis/QGIS#49667]
### What is the bug or the crash?

I am running Ubuntu 20.04. I have three versions of python: 2.7, 3.8. 3.9.

My psycopg2 is installed  in python3.9/site-packages. I cannot find a way to install it in python3.8/site-packages.

Using either python 3.8 or 3.9, I get the error message "No module named psycopg2". If I copy the entire psycopg2 directory from python3.9/site-packages to python3.8/site-packages, I get the message "No module named psycopg2._psycopg". Setting PYTHONPATH has no effect. Attempting import psycopg2 from a python3.9 prompt works, but then qgis will not import. Attempting to import psycopg2 from a python3.8 prompt does not work, but qgis works.

Attempting to load psycopg2 from pip results in the error "No module named 'pip._internal.vcs.mercurial'"; this bug has been around for a long time and has been reported. No action has been taken on it.

### Steps to reproduce the issue

Replicate my directory structure (site3.9/dist-pacakges/psycopg2), python3.8, python3.9, apt-get installation of pycopg2.

Run qgis.bin.

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

QGIS version | 3.10.4-A Coruña | QGIS code branch | Release QGIS code branch.3
-- | -- | -- | --
Compiled against Qt | 5.12.8 | Running against Qt | 5.12.8
Compiled against GDAL/OGR | 3.0.4 | Running against GDAL/OGR | 3.0.4
Compiled against GEOS | 3.8.0-CAPI-1.13.1 | Running against GEOS | 3.8.0-CAPI-1.13.1
Compiled against SQLite | 3.31.1 | Running against SQLite | 3.31.1
PostgreSQL Client Version | 12.2 (Ubuntu 12.2-4) | SpatiaLite Version | 4.3.0a
QWT Version | 6.1.4 | QScintilla2 Version | 2.11.2
Compiled against PROJ | 6.3.1 | Running against PROJ | Rel. 6.3.1, February 10th, 2020
OS Version | Ubuntu 20.04.4 LTS
Active python plugins | MetaSearch; db_manager

</body></html>QGIS version
3.10.4-A Coruña
QGIS code branch
[Release QGIS code branch.3](https://github.com/qgis/QGIS/tree/release-QGIS%20code%20branch_3)
Compiled against Qt
5.12.8
Running against Qt
5.12.8
Compiled against GDAL/OGR
3.0.4
Running against GDAL/OGR
3.0.4
Compiled against GEOS
3.8.0-CAPI-1.13.1
Running against GEOS
3.8.0-CAPI-1.13.1 
Compiled against SQLite
3.31.1
Running against SQLite
3.31.1
PostgreSQL Client Version
12.2 (Ubuntu 12.2-4)
SpatiaLite Version
4.3.0a
QWT Version
6.1.4
QScintilla2 Version
2.11.2
Compiled against PROJ
6.3.1
Running against PROJ
Rel. 6.3.1, February 10th, 2020
OS Version
Ubuntu 20.04.4 LTS
Active python plugins
MetaSearch; 
db_manager

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

**Comment by gioman** [comment]
How this is a QGIS bug?

**Comment by fossildoc** [comment]
Giovanni -

Thank you for responding.

Your web site contains download and installation instructions for QGIS. I
followed the instructions exactly, and it did not install. That makes it a
QGIS bug.

Although I have psycopg2 installed and use it regularly with postgresql, I
should not have to worry about it because the QGIS instructions do not list
it, or anything else, as a dependency.

**Comment by gioman** [comment]
> Your web site

"My" web site? Do you mean QGIS's?



> 3.10.4-A Coruña

This is waaaaay too old. Supported releases are now 3.22 and 3.26. Also if you multiple installations of Python you should also be able to handle them the correct way, especially if you build from source software that depends on it. If you are installing from repository it should work fine, but your mileage may very, depending on how those python installation are set up in your system.

**Comment by fossildoc** [comment]
I went to the following web site, which I denote as "your" web site because
you apparently represent QGIS:
sys.path.insert(0,
"/usr/local/lib/python3.9/dist-packages/pyshp-2.3.1-py3.9.egg")

I then scrolled down to the section labeled Debian/Ubuntu Quickstart.
In that section it says: "Here you will simply install the latest stable
QGIS (3.26.x Buenos Aires) in your Debian or Ubuntu without having to edit
config files."

Since 3.26 is what I wanted, I followed all the instructions which appeared
immediately thereafter. I executed the following:

sudo apt install gnupg software-properties-common
wget -qO - https://qgis.org/downloads/qgis-2022.gpg.key | sudo gpg
--no-default-keyring --keyring
gnupg-ring:/etc/apt/trusted.gpg.d/qgis-archive.gpg --importsudo chmod
a+r /etc/apt/trusted.gpg.d/qgis-archive.gpg
sudo add-apt-repository "deb https://qgis.org/ubuntu $(lsb_release -c -s) main"
sudo apt update
sudo apt install qgis qgis-plugin-grass

Your instructions delivered to me version 3.10.4.

All of the text that follows in a section labeled "Repositories" is
confusing. I have never before seen a download site
which did not default to the latest stable version. No doubt you
experience some sense of superiority when you blame me for your
incompetence, but the fact is that you are a small man with a small
brain and your product to doomed to failure.
